staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,595 @@
+		// technique to pick a queue, dequeue the request at the head of that
+		// queue, increment the count of the number executing, and send true to
+		// the request's channel.
+		qs.dequeueWithChannelAsMuchAsPossible()
+		// ------------------------------------------------------------------------
+		return false, false, false, func() {}
+	}()
+	if shouldReturn {
+		return tryAnother, execute, afterExecution
+	}
+
+	// ************************************************************************
+	// Step 3:
+	// After that method finishes its loop and returns, the final step in Wait
+	// is to `select` (wait) on a message from the enqueud packet's channel
+	// and return appropriately.
+	qs.counter.Add(-1)
+
+	select {
+	case execute := <-req.DequeueChannel:
+		if execute {
+			// execute the request
+			return false, true, func() {
+				qs.finishRequestAndDequeueWithChannelAsMuchAsPossible(req)
+			}
+		}
+		klog.V(5).Infof("request timed out after being enqueued\n")
+		metrics.AddReject(qs.name, "time-out")
+		return false, false, func() {}
+	case <-ctx.Done():
+		klog.V(5).Infof("request cancelled\n")
+		func() {
+			qs.lockAndSyncTime()
+			defer qs.lock.Unlock()
+
+			// TODO(aaron-prindle) add metrics to these two cases
+			if req.Enqueued {
+				// remove the request from the queue as it has timed out
+				for i := range req.Queue.Requests {
+					if req == req.Queue.Requests[i] {
+						// remove the request
+						req.Queue.Requests = append(req.Queue.Requests[:i],
+							req.Queue.Requests[i+1:]...)
+						break
+					}
+				}
+				// At this point, if the qs is quiescing,
+				// has zero requests executing, and has zero requests enqueued
+				// then a call to the EmptyHandler should be forked.
+				if qs.emptyHandler != nil && qs.numRequestsEnqueued == 0 &&
+					qs.getRequestsExecutingLocked() == 0 {
+					qs.counter.Add(1)
+					go func(eh fq.EmptyHandler) {
+						eh.HandleEmpty()
+						qs.counter.Add(-1)
+					}(qs.emptyHandler)
+				}
+			} else {
+				// At this point we know that req was in its queue earlier and another
+				// goroutine has removed req from its queue and called qs.counter.Add(1)
+				// in anticipation of unblocking this goroutine through the other arm of this
+				// select. In this case we need to decrement the counter because this goroutine
+				// was actually unblocked through a different code path.
+				qs.counter.Add(-1)
+			}
+		}()
+		return false, false, func() {}
+	}
+	// ************************************************************************
+}
+
+// syncTimeLocked is used to sync the time of the queueSet by looking at the elapsed
+// time since the last sync and this value based on the 'virtualtime ratio'
+// which scales inversely to the # of active flows
+func (qs *queueSet) syncTimeLocked() {
+	realNow := qs.clock.Now()
+	timesincelast := realNow.Sub(qs.lastRealTime).Seconds()
+	qs.lastRealTime = realNow
+	var virtualTimeRatio float64
+
+	activeQueues := 0
```

Comment:
I didn't intend to ask for you to inline this chunk, sorry if that's how my comments read. The thing that was (and still is) unclear is why this is the right thing to do. I think this chunk was better in its own function and should have its own unit tests.

Reply:
How much of an explanation are you looking for?  There is a discussion at https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#fair-queuing-for-server-requests .  The "min(sum[over q] reqs(q, t), C)" part could be explained further.  This compares to "1" in fair queuing for an output link from a router.  The change from 1 to C is because we are allowed to execute up to C requests at a time, rather than transmit just one packet at a time.  The "min" part is because there is not always enough offered load; when there are fewer than C requests available to execute then the left side of the "min" is the number of requests executing.

Reply:
It's a lot more palatable now that I see that the ` / float64(activeQueues)` is outside of the min call and not inside it.

Reply:
A unit test demonstrating the expected properties of this function might be a good TODO.